src/app/modules/datagrid/datagrid-controller.js

```diff
@@ -20,116 +21,51 @@
 	angular.module('BloomLibraryApp.datagrid')
 	.controller('DataGridCtrl', ['$scope', '$timeout', 'bookService', '$state', '$stateParams', '$location',
 		function ($scope, $timeout, bookService, $state, $stateParams, $location) {
+			$scope.getBooks = function() {
+				var first = 0;
+				var count = bookService.getAllBooksCount();
 
-			$scope.filterOptions = {
-				filterText: "", // gets updated by user action in ngGrid
-				useExternalFilter: true
-			};
-			$scope.totalServerItems = 0;
-			$scope.pagingOptions = {
-				pageSizes: [20, 50, 100],
-				pageSize: 20, // Gets updated by user action in ngGrid
-				currentPage: 1 // Gets updated by user action in ngGrid
-			};
-			$scope.sortInfo = {
-				fields: [], directions: [], columns: []
-			};
-			$scope.getBookRange = function (count, currentPage, searchText) {
-				var first = (currentPage - 1) * count;
-				var sortField = $scope.sortInfo.fields[0];
-				var sortBy = null;
-				// Todo: setting sortBy to a complex field like this does not work...no sorting happens.
-				// It appears we will need to put a redundant top-level data field in the record for
-				// anything we want to sort by. This is especially annoying in that it may not actually
-				// be useful to sort by all the fields...but ng-grid allows the user to do so.
-				// Possibly we could disallow some fields with a message and switch to some other sort.
-				// Another reason a redundant field is probably necessary is that parse.com's built-in sorting
-				// is case-sensitive, and their only suggestion for overcoming this is a redundant
-				// all-lower-case field. Similarly any other sort-key field must somehow be in a form
-				// where a plain binary sort will give the right results.
-				if (sortField == 'title') {
-					sortBy = 'volumeInfo.title';
-				}
-				bookService.getFilteredBookRange(first, count, searchText, "", "", "", sortBy).then(function (result) {
-					$scope.visibleBooks = result;
-					$scope.visibleData = $scope.visibleBooks.map(function (item) {
+				bookService.getFilteredBookRange(first, count, '', '', '', '', '', '', true).then(function (result) {
```

Comment:
Yes, I found the same thing when I did the old version. It would have been MUCH easier to grab them all and let the control handle it. It's one reason (IIRC) I only got filters implemented partially, if at all.
It's theoretically possible to make parse return multiple groups of 1000 records, concatenate the results, and give them to the ui-grid. That would only scale so far, of course, before we take too long or run short of memory or something. But it might be enough, especially if Suzanne continues to be the only contributor to the library. I rather hope it would prove to be a problem, though.
We could also argue that the user is unlikely to want to scroll through 1000+ results, and just give them the first 1000 that match their filter, with some indication that there are more...you could run that idea by JohnH.
At one point JohnH created a sample database with 4000 or so records to get more idea how things would scale. He might still have the code he used...I think he copied from some public repository of English books.
If it proves to be a huge hassle to get it working without downloading everything at once, I suggest exploring with JohnH the option of only showing the first N (where N is however many we think useful and find possible to download quickly).
It's a nuisance that we don't HAVE 1000 records so we don't actually run into the limitation. You can, however, set the limit of how many results to return lower; one way to test would be to make the actual server query be limited to say 50 results.